tools/tests/unit/test_spice_examples.py

```diff
@@ -0,0 +1,100 @@
+from pathlib import Path
+
+import numpy as np
+import pytest
+import spiceypy as spice
+
+from tools.spice.spice_examples import (
+    build_annotated_events,
+    get_attitude_timerange,
+    get_particle_velocity,
+)
+from tools.spice.spice_utils import (
+    list_files_with_extensions,
+)
+
+
+@pytest.fixture()
+def kernel_directory():
+    """Kernel directory."""
+    kernel_directory = Path(__file__).parents[1] / "test_data" / "spice"
+    return kernel_directory
+
+
+@pytest.fixture()
+def kernels(kernel_directory):
+    """Loads all kernels."""
+    kernels = list_files_with_extensions(
+        kernel_directory, [".tsc", ".tls", ".tf", ".bsp"]
+    )
+    return kernels
+
+
+@pytest.fixture()
+def direct_events():
+    """Loads direct events test data.
+
+    Possible example of future direct_events structure:
+    direct_events.keys()
+    dict_keys(['eventID', 'tdb', 'xstart', 'ystart', 'xstop', 'ystop',
+    'eventtype', 'ph', 'spinphase', 'tof', 'ishead90', 'yf', 'd',
+    'energy', 'rvecultra', 'vultra', 'rmagultra', 'rnormultra'])
+    """
+    direct_events = {
+        "vultra": np.array(
+            [-5.7718470e08, -8.8437594e08, -9.5545958e08], dtype=np.float32
+        ),
+        "tdb": 802100000.0,
+    }
+
+    return direct_events
+
+
+def test_get_attitude_timerange(kernels, kernel_directory):
+    """Tests get_attitude_timerange function."""
+
+    spice.kclear()
+
+    spice.furnsh(kernels)
+
+    ck_kernel = list_files_with_extensions(kernel_directory, [".ck"])
+
+    expected_error_message = "ID -43 not found in CK file"
+    with pytest.raises(ValueError, match=expected_error_message):
+        get_attitude_timerange(ck_kernel, -43)
+
+    # Load the C kernel
+    spice.furnsh(ck_kernel)
+
+    start, end = get_attitude_timerange(ck_kernel, -43000)
+
+    # Clear all kernels
+    spice.kclear()
+
+    assert start, end == (802094470.1857615, 802105267.1857585)
+
+
+@pytest.mark.xfail(reason="Add de430.bsp to test/test_data/spice")
+def test_get_particle_velocity(direct_events, kernels, kernel_directory):
+    """Tests the get_particle_velocity function."""
+
+    spice.kclear()
+
+    spice.furnsh(kernels)
+    ck_kernel = list_files_with_extensions(kernel_directory, [".ck"])
+    spice.furnsh(ck_kernel)
+
+    get_particle_velocity(direct_events)
+
+    spice.kclear()
```

Comment:
Not for now, and not for here, but we should come up with some design to abstract away this `spice.kclear()` `spice.furnsh` `spice.kclear()` sequence in spice_utils. (perhaps using `with` keyword? https://www.geeksforgeeks.org/with-statement-in-python/)

Reply:
I disagree, I think it is definitely for here and should be updated. This will mess things up if there is an error during the test or we run tests in different orders and things aren't cleaned up properly.

Reply:
For EMUS, we abstracted away having to think too hard about furnishing kernels by writing a decorator that was added to any function making spiceypy calls. Here is the signature and documentation from that decorator:
```
def ensure_spice(f_py: callable = None,
                 time_kernels_only: bool = False,
                 start_time: int = None, end_time: int = None, time_units: str = 'sclk',
                 kernel_type: str = None):
    """
    Before trying to understand this piece of code, read this:
    https://stackoverflow.com/questions/5929107/decorators-with-parameters/60832711#60832711

    Decorator/wrapper that tries to ensure that a metakernel is furnished in as complete a way as possible.

    ======================
    Control flow overview:
    ======================
    1. Try simply calling the wrapped function naively.
    --> SUCCESS? Great! We're done.
    --> SpiceyError? Go to step 2.

    2. Furnish metakernel at EMUS_METAKERNEL_PATH
    --> SUCCESS? Great, return the original function again (so it can be re-run).
    --> KeyError? Seems like EMUS_METAKERNEL_PATH isn't set, no problem. Go to step 3.

    3. Did we get start_time and end_time parameters passed to the wrapper?
    --> YES? Cool, that means the user thinks we need a full metakernel to run this function. Go fetch one and furnish
        and return the original function (so it can be re-run).
    --> NO? There's still another possibility. Go te step 4.

    4. Did we get the parameter time_kernels_only=True?
    --> YES? We only need LSK and SCLK kernels to run this function. Go fetch those and furnish and return the original
        function (so it can be re-run).
    --> NO? Dang. This is sort of the end of the line. Re-raise the error generated from the failed spiceypy function
     call but add a better message to it.

    ======
    Usage:
    ======
    Three ways to use this object:
    1) A decorator with no arguments (will never query SDC API)
    ```
    @ensure_spice
    def my_spicey_func(a, b):
        pass
    ```
    2) A decorator with hardcoded, permanent parameters with which to fetch a metakernel from the API. This is useful
    if we only need the latest SCLK and LSK kernels for the function involved.
    ```
    @ensure_spice(time_kernels_only=True)
    def my_spicey_api_func(a, b):
        pass
    ```
    3) An explicit wrapper function, providing dynamic parameters to the SDC API call.
    ```
    wrapped = ensure_spice(spicey_func, start_et=362098457, end_time=37485342345, kernel_type='final')
    result = wrapped(a, b)
    ```

    :param f_py: callable, required, The function requiring SPICE that we are going to wrap if being used explicitly,
        Otherwise None, in which case ensure_spice is being used, not as a function wrapper (see l2a_processing.py) but
        as a true decorator without an explicit function argument.
    :param time_kernels_only: bool, optional, Specify that we only need to furnish time kernels (if
        EMUS_METAKERNEL_PATH is set, we still just furnish that metakernel.
    :param start_time: int, optional, Start ephemeris time to pass to the SDC SPICE API
    :param end_time: int, optional, End ephemeris time to pass to the SDC SPICE API
    :param time_units: str, optional, Defaults to SCLK seconds but may also be set to 'J2000' (ephemeris time)
    :param kernel_type: str, optional, Force specific kernel type (recon, final) in API call.
    """
```

I would recommend doing something similar for IMAP. I'm not sure where adding this type of feature fits with the current spice objectives. There are certainly some decisions that need to be made regarding the handling of spice kernels and metakernels that need to be make prior to implementation.

Reply:
A decorator seems like a great way to do this, I like it. Plus using shared code/ideas from other projects would be great!